### PR TITLE
fix: configure Hermetic Python

### PR DESCRIPTION
**Pull request by vam-google**
Also integrate latest XLA changes and apply corresponding fixes (`XlaProgram` -> `HloProgram` renaming caused by https://github.com/openxla/xla/pull/12302

### CODE REVIEW

File: codegen/BUILD [hunk 1]
```diff
@@ -8,6 +8,10 @@ py_binary(
         "//torch_xla/csrc:aten_xla_type.cpp",
         "@torch//:torchgen_deps",
     ],
+    deps = [
+        "@pypi_torch//:pkg",
```
Comment (lsy323) [inline codegen/BUILD:12]: Why is this needed? Is this pulling pytorch whl from pypi?

File: WORKSPACE [hunk 2]
```diff
@@ -59,6 +59,32 @@ http_archive(
...
+# Initialize hermetic Python
+load("@xla//third_party/py:python_init_rules.bzl", "python_init_rules")
+
+python_init_rules()
+
+load("@xla//third_party/py:python_init_repositories.bzl", "python_init_repositories")
+
+python_init_repositories(
+    requirements = {
+        "3.10": "//:requirements_lock_3_10.txt",
```
Comment (lsy323) [inline WORKSPACE:71]: We need 3.8 - 3.11 support, as we release whl for all python versions

File: WORKSPACE [hunk 3]
```diff
@@ -39,7 +39,7 @@ python_configure(
 #    curl -L https://github.com/openxla/xla/archive/<git hash>.tar.gz | sha256sum
```
Comment (lsy323) [inline WORKSPACE:39]: Please also update the `data` in setup.py so that other dependencies can match openxla/xla https://github.com/pytorch/xla/pull/7087/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7L67

File: requirements.in [hunk 4]
```diff
@@ -0,0 +1,27 @@
+astunparse
```
Comment (lsy323) [inline requirements.in:1]: Can you please explain is this requirement file is for bazel build or runtime?

**Review by lsy323 — CHANGES_REQUESTED** [review]

**Review by will-cromar — CHANGES_REQUESTED** [review]
Will hermetic Python work for us? We rely on a locally installed package built from the host's Python environment (`torch`/`torchgen`). The `torch` packages that upstream distributes don't contain all of the tools and headers we need. AFAIK `torch` cannot be built consistently within Bazel.

Where do these lock files come from and how can we keep them updated?

File: codegen/BUILD (hunk 1 above)
Comment (vam-google) [reply codegen/BUILD:12]: No, it is using whichever pytorch which is mentioned in requirements_lock.txt file. To support your workflow it will be locally built .whl file. For what its worth, it could pull pytorch from pypi as well, but I don't think it is what you need here.

Note this target has always been depending on pytorch, but non-hermetically (installed on the system) that is why the dependency was not explicitly defined. Now, since it is hermetic the dependency must be explicit (which is  good and idiomatic thing).

File: WORKSPACE (hunk 2 above)
Comment (vam-google) [reply WORKSPACE:71]: Ok, I'll add (in  new brahcn-based PR) 3.8 as well.

File: WORKSPACE (hunk 3 above)
Comment (vam-google) [reply WORKSPACE:39]: Which exact date should it be?

File: requirements.in (hunk 4 above)
Comment (vam-google) [reply requirements.in:1]: yes, it defines hermetic python environment within bazel

File: WORKSPACE (hunk 3 above)
Comment (lsy323) [reply WORKSPACE:39]: The date is used for locating the nightly libtpu whl. libtpu and openxla need to match. The data should be 1 day prior to the pinned xla commit. (e.g. the pinned xla commit in on May 21st, let's put `20240520` here)

File: WORKSPACE (hunk 2 above)
Comment (lsy323) [reply WORKSPACE:71]: Sounds good, thanks!

File: requirements.in (hunk 4 above)
Comment (lsy323) [reply requirements.in:1]: Thanks for the explanation, I think PyTorch/XLA only use bazel to build a shared library [here](https://github.com/pytorch/xla/blob/master/BUILD#L7), bazel test targets also don't have any python dependencies.(Bazel test is only used for cpp tests) Therefore, maybe we can remove all of these? (Or at least something like `google-api-python-client`, `google-cloud-storage`, I guess we definitely don't need those in the bazel build)

**Comment by JackCaoG** [comment]
I think we don't need this pr anymore, let me close this.
